sql: reduce projects, improve span display in EXPLAIN

RaduBerinde opened the pull request:
#### sql: run each logictest on a single config -rewrite mode

The -rewrite mode is very slow because it disables parallelization.

We don't need to rewrite each test multiple times (for each config). This change
skips any tests that were already run under a different config when the
`-rewrite` flag is passed.

Release note: None

#### sql: reduce projects in EXPLAIN

Projections are very simple operations and are not useful when understanding a
plan.

We remove all projections from the non-verbose EXPLAIN.

In EXPLAIN (VERBOSE), we remove projections that only rearrange / rename
columns. We show the rearranged/renamed columns, but without showing the
separate project node. It would be confusing to always remove all projects in this
mode because the output columns and orderings are visible.

Release note (sql change): EXPLAIN now shows fewer "project" nodes.

#### sql: improve span display in EXPLAIN

This change modifies the spans in the non-verbose EXPLAIN. For regular indexes,
we show the "logical" spans (optimizer style), and limit to a small number of
spans. For inverted indexes, we only show the number of spans (these are not
generally user readable).

The VERBOSE variant is unchanged, always showing all the physical spans.

Release note (sql change): improved display of spans in EXPLAIN.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52865)
<!-- Reviewable:end -->

rohany commented on pkg/sql/opt/exec/execbuilder/testdata/interleaved:
```diff
@@ -58,7 +58,7 @@ EXPLAIN SELECT * FROM level4 WHERE k1 > 1 AND k1 < 3
 ·     missing stats  ·
 ·     table          level4@primary
-·     spans          /2/#/54/1/#/55/1-/2/#/54/1/#/55/2
+·     spans          [/2 - /2]
```
I think you should use the verbose spans here, as these are trying to look at the interleaved keys

rohany commented on pkg/sql/opt/exec/explain/emit.go:
```diff
@@ -31,6 +32,14 @@ func Emit(plan *Plan, ob *OutputBuilder, spanFormatFn SpanFormatFn) error {
+		// or rename the columns).
+		if !ob.flags.Verbose {
+			for n.op == serializingProjectOp || n.op == simpleProjectOp {
+				n = n.children[0]
```
just a nit, but I think that `return walk(n.children[0])` is cleaner.

also, this is a pretty hacky way to get recursion :D though it is the common example of how a programming language with only functions and mutable memory can get you turing completeness

rohany commented on pkg/sql/logictest/testdata/logic_test/inverted_filter_geospatial_explain_local:
```diff
@@ -27,39 +27,35 @@ https://cockroachdb.github.io/distsqlplan/decode.html#eJyUk99v2jAQx9_3V1j3UpC8Yj
+           └── scan        ·                ·
+·                          missing stats    ·
+·                          table            geo_table2@geom_index
+·                          spans            20 spans
```
shouldn't this scan show some spans and then ... some more? Or is it because it is an inverted index?

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/inverted_filter_geospatial_explain_local, line 61 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52865#-MEo2FapxNKLdN17rFLV-r3-61:-MExwkIv3ziOXRvJSH8X:bck6ogx) ([raw file](https://github.com/cockroachdb/cockroach/blob/54e5b7ec4411e8dc0ee5b5458860287149a327ed/pkg/sql/logictest/testdata/logic_test/inverted_filter_geospatial_explain_local#L61)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

shouldn't this scan show some spans and then ... some more? Or is it because it is an inverted index?
</blockquote></details>

Yeah we just show the count for inverted index spans (unless you use VERBOSE), since they are not really user-readable.
___
*[pkg/sql/opt/exec/execbuilder/testdata/interleaved, line 61 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52865#-MEo2Fas0cDAuF5RuGHN-r3-61:-MExxIuIAJDTeyaMzxZf:b-4c5ezp) ([raw file](https://github.com/cockroachdb/cockroach/blob/54e5b7ec4411e8dc0ee5b5458860287149a327ed/pkg/sql/opt/exec/execbuilder/testdata/interleaved#L61)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I think you should use the verbose spans here, as these are trying to look at the interleaved keys
</blockquote></details>

Good catch, done.
___
*[pkg/sql/opt/exec/explain/emit.go, line 40 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52865#-MEo2Farn8MwEf3RRou--r3-40:-MExxtKcAEX91IPZpp20:b-smskfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/54e5b7ec4411e8dc0ee5b5458860287149a327ed/pkg/sql/opt/exec/explain/emit.go#L40)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

just a nit, but I think that `return walk(n.children[0])` is cleaner.

also, this is a pretty hacky way to get recursion :D though it is the common example of how a programming language with only functions and mutable memory can get you turing completeness
</blockquote></details>

Done. It was like this because I initially had a more complicated version where we were saving and using the original `n.Ordering()` (but I saved that for later)


<!-- Sent from Reviewable.io -->

rohany commented on pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans:
```diff
@@ -132,270 +132,141 @@ NULL       /0         {1}       1
-query TTT
-EXPLAIN SELECT * FROM p2@p2_id WHERE i >= 1 AND d IS NOT NULL
+query T
+SELECT description FROM [EXPLAIN SELECT * FROM p2@p2_id WHERE i >= 1 AND d IS NOT NULL] WHERE field = 'spans'
```
missing verbose here

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52865)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans, line 244 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/52865#-MEo2Fas0cDAuF5RuGHL-r5-244:-MEy1rqOCXmmg_gBGgIY:b-qqx0w8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d8c1cdb9f5efc6a60d49b8ef41776d3d07193ab/pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans#L244)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

missing verbose here
</blockquote></details>

Done. TFTR!


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde committed:
sql: improve span display in EXPLAIN

This change modifies the spans in the non-verbose EXPLAIN. For regular indexes,
we show the "logical" spans (optimizer style), and limit to a small number of
spans. For inverted indexes, we only show the number of spans (these are not
generally user readable).

The VERBOSE variant is unchanged, always showing all the physical spans.

Release note (sql change): improved display of spans in EXPLAIN.

RaduBerinde committed:
sql: reduce projects in EXPLAIN

Projections are very simple operations and are not useful when understanding a
plan.

We remove all projections from the non-verbose EXPLAIN.

In EXPLAIN (VERBOSE), we remove projections that only rearrange / rename
columns. We show the rearranged/renamed columns, but without showing the
separate project node. It would be confusing to always remove all projects in
this mode because the output columns and orderings are visible.

Release note (sql change): EXPLAIN now shows fewer "project" nodes.

RaduBerinde committed:
sql: run each logictest on a single config -rewrite mode

The -rewrite mode is very slow because it disables parallelization.

We don't need to rewrite each test multiple times (for each config). This change
skips any tests that were already run under a different config when the
`-rewrite` flag is passed.

Release note: None

RaduBerinde commented:
bors r+
